First draft of Normalize plugin

bhushankawadkar opened the pull request:
First draft of Normalize plugin added.
1. Plugin code added as per requirement and design.
2. JUnit test not added as failing due to error Caused by: java.io.IOException: Incomplete document in MockSink. Detail description emailed. Once resolve this issue will add the test file.

bhushankawadkar committed:
First draft of Normalize plugin

bhushankawadkar committed:
Fields validation with input schema added

bhushankawadkar commented:
First draft of Normalize plugin added. (please consider both commit [b70152e](https://github.com/caskdata/hydrator-plugins/pull/327/commits/b70152e6afcb2708db7c003afa2a92a907ecbc31 and [5e7766c]%28https://github.com/caskdata/hydrator-plugins/pull/327/commits/5e7766c6c073d5a4130bcc4bcff234795e83c100%29)
1. Plugin code added as per requirement and design.
2. JUnit test not added as failing due to error Caused by: java.io.IOException: Incomplete document in MockSink. Detail description emailed. Once resolve this issue will add the test file.

nitinmotgi commented on transform-plugins/src/main/java/co/cask/hydrator/plugin/Normalize.java:
```diff
@@ -0,0 +1,170 @@
+    String[] fieldMappingArray = config.fieldMapping.split(",");
+    for (String fieldMapping : fieldMappingArray) {
+      String[] mappings = fieldMapping.split(":");
+      Preconditions.checkArgument(inputSchema.getField(mappings[0]) != null, "Mapping field '" + mappings[0]
```
If mappings is null, wouldn't this line through a exception ?

bhushankawadkar replied:
Tested on hydrator UI, widget does not allow null or empty key. It throws error, hence never be null.

nitinmotgi replied:
You don't need to use Hydrator to create JSON. You can also create a JSON by hand. So, handling that situation is also important.

bhushankawadkar replied:
sure I will update the code wherever required

nitinmotgi commented on transform-plugins/src/main/java/co/cask/hydrator/plugin/Normalize.java:
```diff
@@ -0,0 +1,170 @@
+    String[] fieldNormalizingArray = config.fieldNormalizing.split(",");
+    for (String fieldNormalizing : fieldNormalizingArray) {
+      String[] fields = fieldNormalizing.split(":");
+      Preconditions.checkArgument(inputSchema.getField(fields[0]) != null, "Normalizing field '" + fields[0]
```
Same here.

bhushankawadkar replied:
Tested on hydratory UI, widget does not allow null or empty value, hence cannot be null.

nitinmotgi commented on transform-plugins/src/main/java/co/cask/hydrator/plugin/Normalize.java:
```diff
@@ -0,0 +1,170 @@
+    String[] fieldMappingArray = config.fieldMapping.split(",");
+    for (String fieldMapping : fieldMappingArray) {
+      String[] mappings = fieldMapping.split(":");
+      Preconditions.checkArgument(mappings.length == 2, "Output schema field is missing for mapping field '" +
```
ditto

bhushankawadkar replied:
same here

nitinmotgi commented:
Minor comments other wise LGTM 👍

bhushankawadkar committed:
Normalize plugin test cases added

bhushankawadkar commented:
JUnit test cases added.
Travis build is failing due to ETLTPFSTestRun, so could not reach upto NormalizeTest class.
However it is building successfully on local machine.
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] Hydrator Plugin Collection ......................... SUCCESS [  2.332 s]
[INFO] Hydrator Common .................................... SUCCESS [  4.186 s]
[INFO] Transform Plugins .................................. SUCCESS [01:58 min]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 02:05 min
[INFO] Finished at: 2016-07-21T18:22:03+05:30
[INFO] Final Memory: 80M/825M
[INFO] ------------------------------------------------------------------------

bhushankawadkar committed:
Review comments updated

bhushankawadkar commented:
Review comments updated
1. added validation for null value check of mapping and normalizing fields
2. added test cases for null / empty mapping and normalizing fields

nitinmotgi commented on transform-plugins/src/main/java/co/cask/hydrator/plugin/Normalize.java:
```diff
@@ -0,0 +1,187 @@
+    outputFields.add(Schema.Field.of(validTypeField, Schema.of(Schema.Type.STRING)));
+    outputFields.add(Schema.Field.of(validValueField, Schema.of(Schema.Type.STRING)));
+    outputSchema = Schema.recordOf("outputSchema", outputFields);
+  }
```
@bhushankawadkar we should have a way to generate the schema from FE. Meaning use the service and integration with frontend with button to generate the schema or we should get user to enter output schema.

bhushankawadkar replied:
@nitinmotgi Code updated to get output schema entered by user and validating it against mapping and normalizing fields.

bhushankawadkar committed:
Validate user entered output schema

bhushankawadkar committed:
Constants made private

bhushankawadkar commented:
Review comments updated
1. Code updated to get output schema entered by user and validating it against mapping and normalizing fields.
2. Test cases added to validate output schema.

nitinmotgi commented:
LGTM :+1:
